rasa/core/policies/rule_policy.py

```diff
@@ -564,7 +564,11 @@ def _check_prediction(
         gold_action_name: Text,
         prediction_source: Optional[Text],
     ) -> List[Text]:
-        if not predicted_action_name or predicted_action_name == gold_action_name:
+        if (
+            not predicted_action_name
+            or not prediction_source
```

Comment:
This is new. Could `None` or `""` be a valid prediction source name?

Reply:
Looks like `_should_delete` would've thrown errors if that was `None`, no?

Reply:
yeah exactly, it cannot be `None`. It doesn't seem like it can be the empty string:

https://github.com/RasaHQ/rasa/blob/b8433f14992151dd816028d9473aa58862539bb3/rasa/core/policies/rule_policy.py#L1089

Reply:
Shall we then get rid of the "Optional" instead of catching something that can't happen?

Reply:
I guess my question is : since it's a prediction, in case we can't predict anything, would prediction_source be `None`?

Reply:
so, some places where that can become None (but that None won't reach the function we're looking at here):
- [find_action_from_default_actions](https://github.com/RasaHQ/rasa/blob/b8433f14992151dd816028d9473aa58862539bb3/rasa/core/policies/rule_policy.py#L924) sets that to ´None´, and, if the source is set to None, then also the action name is set to None (and so the return type there should better be something like `Optional[Tuple[...`. Which is the reason why this is not ending up in this function (cause that source is only forwarded if the action name is not None).
- similarly for [_find_action_from_rules](https://github.com/RasaHQ/rasa/blob/b8433f14992151dd816028d9473aa58862539bb3/rasa/core/policies/rule_policy.py#L924) where the source can only be `None` if the corresponding action name is None (and hence, optional tuple would be more appropriate imho). Empty string is for initialisation only and also won't happen. 
- ...

So my conclusions are:
1. we could improve typing to express that logic better (i.e. source is specified if and only if name is specified) but that requires more changes
2. It definitely cannot be None because otherwise `should_delete` would fail [here](https://github.com/RasaHQ/rasa/blob/b8433f14992151dd816028d9473aa58862539bb3/rasa/core/policies/rule_policy.py#L554) (well, unless all rules are empty... ?)
3. 1+2 are annoying and hence sorry for the spam + let's keep your workaround + let's add a comment explaining why this was added to avoid confusion in the future :D

Reply:
thanks for detailing, I'll add a comment

Reply:
I've quickly tried to changed the return type of the 2 functions you mentioned using `Union[...]` but it does require some refactoring in the code. So the comment is definitely the way to go in this PR :)